### PR TITLE
Use platformio's extend feature

### PR DESCRIPTION
**Pull request by TheAssassin**
While working on Marlin, I found a new (well, new to me) feature in PlatformIO. Using `extends` saves duplicate code and helps maintain similar environments (e.g., the ones using the same platform, framework etc.). I've modified the PlatformIO configuration to use those. I also took the moment and improved the comments. (I *hate* any form of inline comments and lines which exceed my screen size...)

### CODE REVIEW

**Comment by probonopd** [comment]
45 to 66 lines, I find it more complicated than before...

**Comment by carl1961** [comment]
using @TheAssassin   platformio.ini i was able to  compile probonopd on windows 10 using vscode with platformio installed. Awesome work @TheAssassin  and @probonopd    The probonopd has really progressed a lot lately

**Comment by probonopd** [comment]
The kudos for PlatformIO go to @TheAssassin really. :+1:  I suspected that PlatformIO was better than what we had before but always was afraid of doing the work in the fear of breaking something.

**Comment by TheAssassin** [comment]
> 45 to 66 lines, I find it more complicated than before...

Mainly because I extended the documentation. At least 4 lines have been added since I moved your inline comment into two separate lines. Then there's more blank lines. Comparing the lines doesn't make that much sense, therefore.

The actual code is less duplicated than before and should avoid forgetting to change single board configurations while editing something in the others. Even if there's 3-4 more lines, it's still *better code*.

Also, adding *new* boards is a lot easier than before. In PlatformIO, there's a definition for most boards out there. Especially in the ESP32 world, there's sooo many boards now, with different flash methods, some via UART, some via other methods. With the STM32 it's similar, they ship a dedicated debug port, but there's a few UART bootloaders, too, all that can be configured on a per-board basis in `platformio.ini`.

By the way: the PR failed because `transfer.sh` timed out. It built just fine.

File: platformio.ini [hunk 1]
```diff
@@ -22,24 +31,37 @@ lib_deps =
...
 board = d1_mini
-framework = arduino
-lib_deps =
-    ${common.lib_deps}
+extends = env:esp8266_base
 
+# esp32dev works for the majority of ESP32 based dev boards
+# there are more specific board configurations available, feel free to send PRs adding new ones
 [env:esp32dev]
-platform = espressif32@1.8.0 ; Corresponds to https://github.com/platformio/platform-espressif32/releases/tag/v1.8.0 = https://github.com/espressif/arduino-esp32/releases/tag/1.0.2
```
Comment (probonopd) [inline platformio.ini]: Why was this pinning removed, by mistake?

Comment (TheAssassin) [reply platformio.ini]: Check the `extends` line. It points you to the environment `env:esp32_base`, where this line is contained. What's missing is the `@1.8.0` there. I'll update the PR.

Comment (TheAssassin) [reply platformio.ini]: Actually, the `@1.8.0` was wrong anyway. It should've been `1.0.2`

Comment (TheAssassin) [reply platformio.ini]: Strike that, `1.8.0` is the right one.

**Comment by TheAssassin** [comment]
I also simplified the PR a bit, removing the need for the `default_envs`. Should be more satisfying now.

**Comment by probonopd** [comment]
Thanks @TheAssassin, ready for merge?

**Comment by TheAssassin** [comment]
Yes.

**Comment by probonopd** [comment]
Thank you @TheAssassin

**Pull request merged**